### PR TITLE
Tests for array::intersects()

### PR DESCRIPTION
**Pull request by atombender**

### CODE REVIEW

File: test/function/array-intersects.yml [hunk 1]
```diff
@@ -0,0 +1,125 @@
+name: "array::intersects() function"
```
Comment (judofyr) [inline test/function/array-intersects.yml:1]:
```suggestion
name: "array::intersects() function"
version: ">= 1.2"
```

This needs a version for the test runners to actually run this test.

File: test/misc/subqueries.yml [hunk 2]
```diff
@@ -15,7 +15,7 @@ documents:
 
 
 tests:
-  - name: "sunquery-union"
```
Comment (judofyr) [inline test/misc/subqueries.yml:18]: 🌞

File: test/function/array-intersects.yml [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+  # Object
+  - variables: {expr: '[{"a": 42}]'}
+  # Geo
+  - variables: {expr: '[geo({"lat": 1, "lng": 1})]'}
+
+- name: Array right
+  query: |
+    array::intersects(~arr~, ~lit~)
+  variables:
+    lit: '[1, "1", null, true, false, dateTime("2002-10-02T12:34:56Z"), {"a": 42}, geo({"lat": 1, "lng": 1})]'
```
Comment (judofyr) [inline test/function/array-intersects.yml:52]:
```suggestion
    lit: '[1, "1", null, true, false, {"a": 42}]'
```

File: test/function/array-intersects.yml [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+  tests:
+  # Equal
+  - result: true
+    variables:
+      arr: '[1, "1", null, true, false, dateTime("2002-10-02T12:34:56Z"), {"a": 42}, geo({"lat": 1, "lng": 1})]'
```
Comment (judofyr) [inline test/function/array-intersects.yml:57]:
```suggestion
      arr: '[1, "1", null, true, false, {"a": 42}]'
```

File: test/function/array-intersects.yml [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  # Superset
+  - result: true
+    variables:
+      arr: '[1, "1", null, true, false, dateTime("2002-10-02T12:34:56Z"), {"a": 42}, geo({"lat": 1, "lng": 1}), "2"]'
```
Comment (judofyr) [inline test/function/array-intersects.yml:61]:
```suggestion
      arr: '[1, "1", null, true, false, {"a": 42}, "2"]'
```

File: test/function/array-intersects.yml [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+      arr:
+      - '[2]'
+      - '["2"]'
+
+# Exactly the same as previous, except reversed order of arguments
+- name: Array left
+  query: |
+    array::intersects(~lit~, ~arr~)
+  variables:
+    lit: '[1, "1", null, true, false, dateTime("2002-10-02T12:34:56Z"), {"a": 42}, geo({"lat": 1, "lng": 1})]'
```
Comment (judofyr) [inline test/function/array-intersects.yml:92]:
```suggestion
    lit: '[1, "1", null, true, false, {"a": 42}]'
```

File: test/function/array-intersects.yml [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+  tests:
+  # Equal
+  - result: true
+    variables:
+      arr: '[1, "1", null, true, false, dateTime("2002-10-02T12:34:56Z"), {"a": 42}, geo({"lat": 1, "lng": 1})]'
```
Comment (judofyr) [inline test/function/array-intersects.yml:97]:
```suggestion
      arr: '[1, "1", null, true, false, {"a": 42}]'
```

File: test/function/array-intersects.yml [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+  # Superset
+  - result: true
+    variables:
+      arr: '[1, "1", null, true, false, dateTime("2002-10-02T12:34:56Z"), {"a": 42}, geo({"lat": 1, "lng": 1}), "2"]'
```
Comment (judofyr) [inline test/function/array-intersects.yml:101]:
```suggestion
      arr: '[1, "1", null, true, false, {"a": 42}, "2"]'
```

**Review by judofyr — CHANGES_REQUESTED** [review]
Some of the variables here are not valid JSON (since they include `dateTime`/`geo`) and this means that the built-in [Automatic test generation based on variables](https://github.com/sanity-io/groq-test-suite#automatic-test-generation-based-on-variables) doesn't kick in. With the changes proposed here we get a bunch of more tests generated.

---

As an aside, we really should re-think the whole "if a variable is JSON then parse it" thingie. It's probably a lot clearer if we do the opposite: The test case contains the JSON, and then we instead serialize it into the query:

```yml
generate:
  variables:
    expr: [1, "1", null, true, false, {"a": 42}]
```

This could then also assert that there is no `documents` (which is another gotcha!)

I think also @rneatherway was a bit surprised as to how the same variable duplicated in a query behaved.

**Comment by atombender** [comment]
Any comment on the above?

**Comment by atombender** [comment]
@judofyr Another bump. 🙂

**Comment by judofyr** [comment]
Separate `dateTime` / `geo` tests sounds perfectly fine to me. They are quite different beasts. We're already lacking tests which checks that `array::intersects` with the _same_ `dateTime` / `geo` returns `false`.

**Comment by atombender** [comment]
> They are quite different beasts. We're already lacking tests which checks that `array::intersects` with the _same_ `dateTime` / `geo` returns `false`.

Note sure what you mea there, since datetimes are [considered primitives](https://github.com/sanity-io/gradient/blob/2d01bc955e043e6725902886a7afea7185f03c61/pkg/search/datamodel/key.go#L26-L27) and are equivalent in terms of comparison in these tests. Sounds like the only problem is the automatic test generation?

For geo not intersecting with itself, there is already a [test for that](https://github.com/sanity-io/groq-test-suite/pull/80/files#diff-9281d464e6eb3b49a5eb063cdc52d17e81f1f2900e73d1358396115e0bf7e1c0R47).

**Review by j33ty — APPROVED** [review]
LGTM

**Pull request merged**